core/commons/che-core-commons-annotations/src/main/java/org/eclipse/che/commons/annotation/Traced.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.commons.annotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Annotates a method as traced.
+ *
+ * <p>If the method is declared in a Guice-managed class and the Guice environment is equipped with
+ * an interceptor for handling this annotation (which should be the case at least in the workspace
+ * server), each call of such method will create a new span declared as a child of some current
+ * span, if any, with the provided name (or the default name).
+ *
+ * <p>The method can declare additional tags in its body that will be applied to the span, see
+ * {@link Tags}.
+ */
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Traced {
+
+  /**
+   * The name of the span generated for the method. Defaults to the "classSimpleName#methodName".
+   */
+  String name() default "";
+
+  /**
+   * Use this class to statically set the tags on the span generated for a {@code @Traced} method.
+   * Note that the tags are applied <b>after</b> the method is invoked no matter where in the method
+   * they are set.
+   */
+  final class Tags {
+
+    /** Adds a new tag. If the tag already exists, it is NOT updated. */
+    public static void add(String tagName, String value) {
+      Map<String, String> tags = TagsStack.TAGS.get().peek();
+      if (tags != null) {
+        tags.putIfAbsent(tagName, value);
+      }
+    }
+
+    /** Adds new tags. The values of existing tags are not updated. */
+    public static void addAll(Map<String, String> keyValues) {
+      Map<String, String> map = TagsStack.TAGS.get().peek();
+      if (map != null) {
+        for (Map.Entry<String, String> e : keyValues.entrySet()) {
+          map.putIfAbsent(e.getKey(), e.getValue());
```

Comment:
Can't we just use `map.putAll(keyValues())`?

Reply:
That would have a different behavior than `putIfAbsent`. Unfortunately there is no `putAllIfAbsent` method I could use to preserve the same behavior as the `add()` method.

Reply:
yeah, you are right. Not sure we need to have absolutely the same behavior, but I agree that it is safer